src/transformers/generation/logits_process.py

```diff
@@ -1878,6 +1890,61 @@ def __call__(self, input_ids: torch.LongTensor, scores: torch.FloatTensor) -> to
         return scores
 
 
+class WhisperNoSpeechDetection(LogitsProcessor):
```

Comment:
You are a 100% correct that this logit processor is quite akward! I'm also not super about the way it's implemented, but for me it was a hard requirement that Whisper doesn't change the generate method at all. 

The problem with retrieving the "no speech" prob from the `scores` is that these scores will be the ones that went through all the logit processors (see [here](https://github.com/huggingface/transformers/blob/6c78bbcb8320d316434262ef003251ca997db0d1/src/transformers/generation/utils.py#L2605)) which will mean that the prob retrieved will necessarily be incorrect. In the original codebase the "no speech" prob is computed from the "untouched" logits (see [here](https://github.com/openai/whisper/blob/ba3f3cd54b0e5b8ce1ab3de13e32122d0d5f98ab/whisper/decoding.py#L693)) 

Other solutions that I have considered here:
- 1.) Just not add the "NoSpeechDetection" logit processor since it doesn't help that much for performance in my tests. 
- 2.) Make the logit processor cleaner by just checking the "scores" that are forwarded to the logit processor. **However** this would require [this line](https://github.com/huggingface/transformers/blob/6c78bbcb8320d316434262ef003251ca997db0d1/src/transformers/generation/utils.py#L2602)to be changed so that all logits (not just the last one are passed into the logit processors). This would though mean I'd have to also add some hacky code to the "LogitsProcessor" class that makes sure that the logits are cut to the last in case `WhisperNoSpeechDetection` is not part of the processors.

=> Overall I think neither 1) nor 2) is better than what we have now. I'm willing to go for 1.) though if people prefer (also cc @sanchit-gandhi)

Reply:
> it was a hard requirement that Whisper doesn't change the generate method at all

100% agreed 👍 

> The problem with retrieving the "no speech" prob from the scores is that these scores will be the ones that went through all the logit processors (see [here](https://github.com/huggingface/transformers/blob/6c78bbcb8320d316434262ef003251ca997db0d1/src/transformers/generation/utils.py#L2605)) which will mean that the prob retrieved will necessarily be incorrect.

I see 😞 Our design makes this feature quite ugly to implement, but I think it's more important to match the reference paper/implementation!